### ISSUE TITLE
Parameterized stub endpoints

### ISSUE THREAD

**Issue by momo-sa** [ithaka/apiron#31]
When working with stubs, we came across circumstances in which it would have been nice to have parameterized stub data. Put another way, we needed to stub out a single endpoint that could yield distinct responses, depending on the input params. The current implementation can only yield a single response, regardless of input params.

[endpoint/stub.py](https://github.com/ithaka/apiron/blob/dev/apiron/endpoint/stub.py)

### Option 1: allow `stub_response` as a function

Thinking about how to do parameterize stubs without breaking existing functionality, we might allow the `stub_response` to be a function taking the would-be endpoint params as input.

Sample `stub_response` we would pass into the endpoint for parameterization:

```python
def stub_response(**params):
    data = {
        'foo': {'bar': 1},
        'default': {'baz': 2},
    }
    data_key = params.setdefault('param_name', 'default')
    return data[data_key]
```

So if `stub_response` is a function, we would return `stub_response(**self.endpoint_params)`; otherwise, proceed as before.

The upshots are that this change wouldn't break current functionality while allowing us to create stub endpoints that adapt to different inputs as needed. The downside is that the `stub_response` gets a little polymorphic, and the name "stub_response" doesn't make it obvious that it could also be a _stub response determiner_.

### Option 2: add an optional parameter for a determiner function, like `response_determiner`

This also doesn't break existing functionality, allowing the user to specify a `stub_response` as before, but now also an optional `response_determiner` function. I'd expect the determiner function to take `**self.endpoint_params` as input as before and yield the desired response data. When the determiner function is provided, `stub_response` isn't used.

The upshots are that we retain current functionality. The downside is that the relationship between `stub_response` and the `response_determiner` may be confusing. Will a user expect the determiner to use `stub_response` in some way, like a data set to select from?

----

IMO the second option is too confusing, so I lean towards Option 1 although it's not perfect. But I'm open to other ideas.

attn: @daneah

**Comment by daneah** [comment]
I'd be inclined toward Option 1 as well, especially considering that `stub` could be taken as a noun or a verb and so covers each case in question. With ample documentation I imagine it would be the less confusing of the two suggestions 👍

**Comment by momo-sa** [comment]
Cool beans @daneah . I'll plan to submit a PR within a week.

**Comment by daneah** [comment]
Resolved in #32
